### PR TITLE
[ie/abematv]: fix extraction when using cache

### LINKED ISSUE #6532: AbemaTV-license
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that a **supported** site is broken
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Japan

### Provide a description that is worded well enough to be understood

When I try to download a video from AbemaTV, an error occurs.
I have added the -u and -p parameters as this video must be viewed with a premium account, I have run it using cookies and the situation is the same.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['--verbose', '-u', 'PRIVATE', '-p', 'PRIVATE', 'https://abema.tv/video/episode/19-149_s1_p4']
[debug] Encodings: locale cp932, fs utf-8, pref cp932, out cp932 (No VT), error cp932 (No VT), screen cp932 (No VT)
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-7-6.1.7601-SP1 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 5.1-full_build-www.gyan.dev (setts), ffprobe 5.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[debug] Loading abematv.mymail@outlook.jp from cache
[AbemaTV] Fetching media token
[debug] Loading abematv.mymail@outlook.jp from cache
[debug] [AbemaTV] Skipping logging in
[AbemaTV] Extracting URL: https://abema.tv/video/episode/19-149_s1_p4
[AbemaTV] 19-149_s1_p4: Downloading webpage
[AbemaTV] 19-149_s1_p4: Checking playability
WARNING: [AbemaTV] This is a premium-only stream
[AbemaTV] 19-149_s1_p4: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 19-149_s1_p4: Downloading 1 format(s): 5300
[debug] Invoking hlsnative downloader on "https://vod-abematv.akamaized.net/program/19-149_s1_p4/1080/playlist.m3u8?aver=1"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 240
[download] Destination: #04 まひろとあたらしい友達 [19-149_s1_p4].mp4
[debug] File locking is not supported. Proceeding without locking
[download] #04 まひろとあたらしい友達 [19-149_s1_p4].mp4.part-Frag1 has already been downloaded

[download]   0.4% of ~ 774.84MiB at  807.14MiB/s ETA Unknown (frag 1/240)ERROR: unable to download video data: <urlopen error unknown url type: abematv-license>
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 3247, in process_info
  File "yt_dlp\YoutubeDL.py", line 2970, in dl
  File "yt_dlp\downloader\common.py", line 444, in download
  File "yt_dlp\downloader\hls.py", line 365, in real_download
  File "yt_dlp\downloader\fragment.py", line 523, in download_and_append_fragments
  File "yt_dlp\downloader\fragment.py", line 364, in decrypt_fragment
  File "yt_dlp\downloader\fragment.py", line 353, in _get_key
  File "yt_dlp\YoutubeDL.py", line 3742, in urlopen
  File "urllib\request.py", line 525, in open
  File "urllib\request.py", line 547, in _open
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 1425, in unknown_open
urllib.error.URLError: <urlopen error unknown url type: abematv-license>
```

### PR DESCRIPTION
**Pull request by sefidel**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

The extractor fails with
```
yt_dlp.networking.exceptions.NoSupportingHandlers: Unable to handle request: Unsupported url scheme: "abematv-license" (requests, urllib)
```
and
```
  File "/home/sefidel/projects/contrib/yt-dlp-fork/yt_dlp/extractor/abematv.py", line 79, in _get_videokey_from_ticket
    (license_response['cid'] + self.ie._DEVICE_ID).encode('utf-8'),
     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
TypeError: can only concatenate str (not "NoneType") to str
```

because when `yt-dlp` fetches the token from cache, it triggers an early return.
Which resulted in `add_opener` not being called, and `_DEVICE_ID` being unset.

This commit fixes the issue by also calling the `add_opener` function when fetching the token from cache, and setting the `_DEVICE_ID` variable earlier.

p.s. I was unable to test it without using the cache (i.e. via username and password auth) since the request kept failing with `409 Conflict`, but I think that's out of the scope of this PR.

Fixes #6532

Note: test is failing, but also fails on latest release.
```
_______________________________________________________________ TestDownload.test_AbemaTV _______________________________________________________________
test/test_download.py:113: in test_template
    raise Exception(f'Test {tname} definition incorrect - "ext" key must be present to define the output file')
E   Exception: Test test_AbemaTV definition incorrect - "ext" key must be present to define the output file
================================================================ short test summary info ================================================================
FAILED test/test_download.py::TestDownload::test_AbemaTV - Exception: Test test_AbemaTV definition incorrect - "ext" key must be present to define the output file
```
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/abematv.py
Comment (pukkandan) [inline yt_dlp/extractor/abematv.py:141]: Why not just move the opener also up here?

```suggestion
        AbemaTVBaseIE._DEVICE_ID = str(uuid.uuid4())
        add_opener(self._downloader, AbemaLicenseHandler(self))

        username, _ = self._get_login_info()
        AbemaTVBaseIE._USERTOKEN = username and self.cache.load(self._NETRC_MACHINE, username)
```

Comment (sefidel) [reply yt_dlp/extractor/abematv.py:141]: Oops, I initially did that and forgot to switch it back after fiddling with the codebase.

Comment (sefidel) [reply yt_dlp/extractor/abematv.py]: Since the suggestion did not remove the `add_opener` call at the end of the function, I've manually applied the changes.

**Review by pukkandan — APPROVED** [review]

**Comment by sefidel** [comment]
This PR has been converted to draft as the current implementation suffers a problem in decryption process with the license key.

To work around this problem, we would have to also store the `_DEVICE_ID` in the cache.
I'll follow up with a commit that contains a fix to this problem.

**Comment by pukkandan** [comment]
Make sure to invalidate old cache using `min_ver` argument of `Cache.load`

**Comment by sefidel** [comment]
What should the value be? I'm pretty sure that there will be a nightly release today, but I'm not sure if I can just set it to 2024.01.01.

**Comment by pukkandan** [comment]
Sure. A maintainer can update it before merge

**Comment by pukkandan** [comment]
Pls avoid force-pushing. You can take as many commits as you need

File: yt_dlp/extractor/abematv.py
Comment (pukkandan) [inline yt_dlp/extractor/abematv.py:142]: note to maintainers: Update the version before merge

**Comment by sefidel** [comment]
Oops, my bad. It wasn't a squash, just edited commits to include some description.

File: yt_dlp/extractor/abematv.py
Comment (pukkandan) [inline yt_dlp/extractor/abematv.py:261]: same

**Comment by Novas8** [comment]
https://abema.tv/video/episode/661-1_s1_p20

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/abematv.py
Comment (bashonly) [reply yt_dlp/extractor/abematv.py:142]:
```suggestion
        auth_cache = username and self.cache.load(self._NETRC_MACHINE, username, min_ver='2024.01.19')
```

File: yt_dlp/extractor/abematv.py
Comment (bashonly) [reply yt_dlp/extractor/abematv.py:261]:
```suggestion
        if self.cache.load(self._NETRC_MACHINE, username, min_ver='2024.01.19') and self._get_media_token():
```

**Comment by Novas8** [comment]
Yes

**Pull request merged**
